man/snapconv.1
.\" -*- nroff -*-
.\"
.\" snapconv.1: snapconv man page
.\" Copyright (c) 2003-2004 Philip Kendall
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: philip-fuse@shadowmagic.org.uk
.\"
.\"
.TH snapconv 1 "24th May, 2013" "Version 1.1.1" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
snapconv \(em Sinclair ZX Spectrum snapshot converter
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B snapconv
[ \-c ] [ \-n ] [ \-f ]
.I infile outfile
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
snapconv converts between ZX Spectrum snapshot files.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.I \-c
specifies that all data in the output snapshot should be compressed,
even if that compression will result in a bigger file than if some
data is left uncompressed. This helps compatibility with some buggy
programs that can only read entirely compressed .z80 files.
.TP
.I \-n
specifies that the output snapshot should be uncompressed.
.TP
.I \-f
specifies that the A and F and A\(aq and F\(aq registers should be swapped.
Version 0.5.0 and older of 
.IR libspectrum "(3)"
had a bug that resulted in these registers being swapped in SZX 
snapshots only. Version 0.5.1 and newer of
.IR libspectrum "(3)"
automatically recognise these snapshots and correct the problem, but
this flag is necessary to fix some snapshots created by
.I development versions only
of
.IR libspectrum "(3)"
which had the bug fixed, but cannot be distinguished from snapshots
containing the bug.
.TP
.I infile
specifies the snapshot to be converted. This file can be in any of the
snapshot formats supported by
.IR libspectrum "(3)."
.TP
.I outfile
specifies the output snapshot. The format will be determined by the
file extension and can be in any of the formats
.IR libspectrum "(3)"
supports for output.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
None known.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR libspectrum "(3),"
.IR listbasic "(1),"
.IR snap2tzx "(1)"
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Philip Kendall (philip\-fuse@shadowmagic.org.uk).


ChangeLog
2013-05-xx  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.1 released.

        * Decreased libtool version to allow binaries built against
          libspectrum 1.0.0 to run with libspectrum 1.1.1.

2013-05-18  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.0 released.

        * Support for Spectranet in snapshots (Philip Kendall).

        * Support for Cheetah SpecDrum in snapshots (Fredrick Meunier).

        * Support reading PZX files (Fredrick Meunier).

        * Support for the TZX set signal level block (Fredrick Meunier).

        * Various minor bug fixes/improvements:
          * Add support for detecting POK files (Sergio Baldoví).
          * Advertise support for bzip2 compressed and wav files (Sergio
            Baldoví).
          * Drop support for GLib version 1.x (Alberto Garcia).
          * Add support for audiofile 0.3.x (Alberto Garcia).
          * Load all snapshot RAM pages for machines with more than 128k 
            (Fredrick Meunier; thanks, ketmar).
          * Support more SZX 1.5 features (Fredrick Meunier).
          * Modernise autoconf support (Sergio Baldoví).

2010-12-16  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.0.0 released.

        * New end of tape flag (Fredrick Meunier).

        * AF and AF' were swapped when reading and writing SZX files
          (Fredrick Meunier; thanks, velesoft).

        * Support for Fuller Box and Melodik (and similar AY interfaces) in
          snapshots (Fredrick Meunier).

        * Support for Opus Discovery in snapshots (Fredrick Meunier, Gergely
          Szasz).

        * Support for the NTSC 48K Spectrum (Philip Kendall).

        * Support for stereo sound files (Fredrick Meunier; thanks, Gilberto
          Almeida).

        * Add proper support for +2A/+3, Pentagon and Scorpion in .z80 files
          (Philip Kendall; thanks, Nick Corey)

        * Various fixes for compiling with Visual Studio 6 (Simon Owen).

        * Various minor bug fixes/improvements:
          * Looser signature checks for .dsk files (Gergely Szasz; thanks,
            zx81 and Simon Owen).
          * --without-(bzip2|zlib) options to configure for Gentoo and similar
            distributions (José Manuel Ferrer Ortiz).
          * Handle querying the block length for turbo and pure data blocks
            with 0 length (Fredrick Meunier).
          * Fix loading of Warajevo TAPs with raw samples (Fredrick Meunier).
          * Fail sensibly if file decompression fails (Fredrick Meunier).
          * Remove various unused variables, initialisations etc (Fredrick
            Meunier).

2008-11-30  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 0.5.0 released.

	* Memory management change: memory allocation functions now abort
	  on failure. Leads to simpler code both internally and for
	  users of the library (Philip Kendall).

	* Support for tape acceleration (Philip Kendall)

	* IDE improvements to allow DivIDE to support FATware (Philip
	  Kendall; thanks, http://www.dusky.sk/zxs/)

	* New API to deal with flagging automatic snapshots in RZX files
	  (Philip Kendall).

	* New function to get length (time) of a tape block (Philip Kendall,
	  Fredrick Meunier).

	* 128K SNA files are now treated as being for the Pentagon 128
	  rather than the Spectrum 128 (Fredrick Meunier)

	* SNA files now written out correctly (Philip Kendall).

	* Handle MDR write protection flag correctly (Philip Kendall)

	* Allow MDR images with bad checksums to load (the error will
	  be handled by the If1 ROM) (Gergely Szasz)

        * Support saving custom ROMs in SZX snapshots (Fredrick Meunier,
	  Stuart Brady)

	* Support for DivIDE, Kempston mouse and the simple IDE interface
	  in SZX snapshots (Fredrick Meunier)

        * Allow building without libgcrypt (Stuart Brady)

	* Amiga and MorphOS compilation fixes (Chris Young, Q-Master)

	* Don't produce empty libraries to fix build process on
	  OpenSolaris (Fredrick Meunier)

	* Remove deprecated functions (Fredrick Meunier).

	* Tweak timings for TS2068 (Fredrick Meunier; thanks Philip Kendall).

        * NTSC Spectrum support (Philip Kendall; thanks, Claudio Bernet).

2008-01-05  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.4.0 released.

	* Add support for loading from WAV files (Fredrick Meunier).

	* Add support for CSW tape files (Fredrick Meunier).

	* Tweak timings for TC2048 and TC2068 (Fredrick Meunier; thanks Philip
	  Kendall and Mark Woodmass).

	* Add support for the +D in z80 and szx snapshots (Fredrick Meunier).

	* Add Pentagon 512/1024 support (Q-Master).

	* Allow Amiga version to decompress files with xfdmaster.library
	  (Chris Young).

	* Allow MorphOS version to use xfdmaster.library (Q-Master).

	* Various other bugfixes.

2007-05-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0.1 released (bug fix release).

	* Avoid crash when reading an invalid TZX file (thanks, crabfists).
	  (Philip Kendall).

	* Export the TZX Generalized Data Block functions when building a
	  Win32 DLL (Stuart Brady).

	* Distribute the hacking/ directory (thanks, Stuart Brady) (Philip
	  Kendall).

	* Remove warning when building from CVS (Stuart Brady).

2007-04-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0 released.

	* Support for RZX rollback, including having snapshots and
	  multiple input recording blocks in input recording files (Philip
	  Kendall).

	* Improved IDE emulation (Matthew Westcott).

	* Support for microdrive images (Philip Kendall, Gergely Szasz).

	* TZX Generalized Data Block support. (Philip Kendall).

	* Support for Z80Em and CSW tape files (Darren Salt).

	* Support for the TS2068 (Fredrick Meunier).

	* Support for connected joystick information in snapshots (Fredrick
	  Meunier).

	* Add interrupt length to timing information (Philip Kendall).


snapconv.c
/* snapconv.c: Convert snapshot formats
   Copyright (c) 2003-2005 Philip Kendall

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   E-mail: philip-fuse@shadowmagic.org.uk

*/

#include <config.h>

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include <libspectrum.h>

#include "compat.h"
#include "utils.h"

char *progname;

int
main( int argc, char **argv )
{
  libspectrum_snap *snap;
  libspectrum_id_t type; libspectrum_class_t class;
  unsigned char *buffer; size_t length;
  libspectrum_creator *creator;
  int flags;
  int compress = 0;
  FILE *f;

  int error;
  int c;

  progname = argv[0];

  while( ( c = getopt( argc, argv, "cn" ) ) != -1 ) {

    switch( c ) {

    case 'c': compress = LIBSPECTRUM_FLAG_SNAPSHOT_ALWAYS_COMPRESS;
      break;

    case 'n': compress = LIBSPECTRUM_FLAG_SNAPSHOT_NO_COMPRESSION;
      break;

    }
  }
  argc -= optind;
  argv += optind;

  if( argc < 2 ) {
    fprintf( stderr, "%s: usage: %s [-c] [-n] <infile> <outfile>\n", progname,
	     progname );
    return 1;
  }

  error = init_libspectrum(); if( error ) return error;

  snap = libspectrum_snap_alloc();

  if( read_file( argv[0], &buffer, &length ) ) {
    libspectrum_snap_free( snap );
    return 1;
  }

  error = libspectrum_snap_read( snap, buffer, length, LIBSPECTRUM_ID_UNKNOWN,
				 argv[0] );
  if( error ) {
    libspectrum_snap_free( snap ); free( buffer );
    return error;
  }

  free( buffer );

  error = libspectrum_identify_file_with_class( &type, &class, argv[1], NULL,
                                                0 );
  if( error ) { libspectrum_snap_free( snap ); return error; }

  if( class != LIBSPECTRUM_CLASS_SNAPSHOT ) {
    fprintf( stderr, "%s: '%s' is not a snapshot file\n", progname, argv[1] );
    libspectrum_snap_free( snap );
    return 1;
  }

  error = get_creator( &creator, "snapconv" );
  if( error ) { libspectrum_snap_free( snap ); return error; }

  length = 0;
  error = libspectrum_snap_write( &buffer, &length, &flags, snap, type,
                                  creator, compress );
  if( error ) {
    libspectrum_creator_free( creator ); libspectrum_snap_free( snap );
    return error;
  }

  if( flags & LIBSPECTRUM_FLAG_SNAPSHOT_MAJOR_INFO_LOSS ) {
    fprintf( stderr,
	     "%s: warning: major information loss during conversion\n",
	     progname );
  } else if( flags & LIBSPECTRUM_FLAG_SNAPSHOT_MINOR_INFO_LOSS ) {
    fprintf( stderr,
	     "%s: warning: minor information loss during conversion\n",
	     progname );
  }
  error = libspectrum_creator_free( creator );
  if( error ) { free( buffer ); libspectrum_snap_free( snap ); return error; }

  error = libspectrum_snap_free( snap );
  if( error ) { free( buffer ); return error; }

  f = fopen( argv[1], "wb" );
  if( !f ) {
    fprintf( stderr, "%s: couldn't open '%s': %s\n", progname, argv[1],
	     strerror( errno ) );
    free( buffer );
    return 1;
  }
    
  if( fwrite( buffer, 1, length, f ) != length ) {
    fprintf( stderr, "%s: error writing to '%s'\n", progname, argv[1] );
    free( buffer );
    return 1;
  }

  free( buffer );
   
  return 0;
}
